Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -803,42 +803,37 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
                     if (!light.is_enabled)
                         continue;
 
-                    // We need to save the length here because the attenuation factor requires a non
-                    // normalized vector!
-                    auto sgi_arrow_operator = [](FloatVector4 const& p1, FloatVector4 const& p2, float& saved_length) {
+                    // We need to save the length here because the attenuation factor requires a non-normalized vector!
+                    auto sgi_arrow_operator = [](FloatVector4 const& p1, FloatVector4 const& p2, float& output_length) {
                         if ((p1.w() != 0.0f) && (p2.w() == 0.0f)) {
-                            saved_length = p2.length();
-                            return (p2 / saved_length).xyz();
+                            output_length = p2.length();
+                            return p2.xyz().normalized();
                         } else if ((p1.w() == 0.0f) && (p2.w() != 0.0f)) {
-                            saved_length = p2.length();
-                            return -(p1 / saved_length).xyz();
+                            output_length = p1.length();
+                            return -p1.xyz().normalized();
                         } else {
                             // FIXME: The OpenGL 1.5 spec says nothing about the case where P1 and P2 BOTH have a w value of 1, which would
                             // then mean the light position has an implicit value of (0, 0, 0, 0). This doesn't make any logical sense, and it most likely
                             // a typographical error. Most other GL implementations seem to just fix it to the distance from the vertex to the light, which
                             // seems to work just fine.
                             // If somebody with more insight about this could clarify this eventually, that'd be great.
-                            auto distance = (p2 - p1);
-                            saved_length = distance.length();
-                            return (distance / saved_length).xyz();
+                            auto distance = p2 - p1;
```

Comment:
I think we need to calculate `p2.xyz() - p1.xyz()` here. Since `w` can be <> 0, doing it this way will give us a wrong value for `length`.

Reply:
Although, this would only be a problem if `p1.w() != p2.w()` which would be weird. 🤔

Reply:
I've refactored that code a bit so we now only calculate the length once.